### ISSUE TITLE
Command to return/use path on host instead of DOSBox-X path?

### ISSUE THREAD

**Issue by emendelson** [joncampbell123/dosbox-x#2602]
It would be very useful (at least it would be useful in my project) to have a command that would let us use the filename on the host instead of the filename in DOSBox-X. For example, when running a host command with START, it would be good to be able to enter something like this:

`START someCommand hostname://C:\Temp\somefile`

The URL hostname://D:\path\file would be replaced by the path on the host. This would make it easy to run (for example) shell scripts with a filename as a parameter.

Also, the command 

`hostname c:\autoexec.bat`

could return the path on the host of the file autoexec.bat.

Maybe this is already in DOSBox-X, but, if so, I haven't found it. it would be a great convenience to have this.

**Comment by Wengier** [comment]
@emendelson The START command passes the parameter as-is to the host-system application, so the parameter(s) of START command already refer to the host system filenames rather than filenames inside DOSBox-X. This also applies to commands like MOUNT, which mounts host-system paths as DOSBox-X drives. As for host-system path of the file autoexec.bat, it is of course not always possible, since the file autoexec.bat may simply reside inside a virtual drive (e.g. Z:\AUTOEXEC.BAT) or in a disk image, rather than on a mounted local drive, and in such cases it can be said that the file does not physically exist in the host system (there are different types of drives in DOSBox-X after all). Hope this helps.

**Comment by Wengier** [comment]
@emendelson Try the updated Windows build, which allows to return the path on the host of the specified file using /H option of TRUENAME command. If the file cannot be found or is not located on a mounted local or overlay drive then it will return empty.

* [dosbox-x-update.zip](http://cndos.fam.cx/dosware/dosbox-x-update.zip)

Hope this helps.

**Comment by emendelson** [comment]
@Wengier - Yes, I understood that the parameter of the START command must be the host-system filename, and I was asking for a way to make it easy to send that host-system filename from (for example) a WordPerfect macro. I was using "autoexec.txt" only as an example. Typically I want to enter a command that runs a shell script on a file, using the host-system path as a parameter for the script. But I have no direct way of knowing what the host-system path of the file might be.

For example, in DOSBox-X, the file might be named `D:\mydocs\myfile.doc`, but it could be anything else.

What I have been doing until now is something like this. First I output a temporary text file with the result of the MOUNT D:\ command and the full DOSBox-X path (`D:\mydocs\myfile.doc`). Then I enter the command:

`START doSomething.sh`

Then the doSomething.sh shell script parses the temporary text file with the result of the MOUNT command the DOSBox-X path to find the host-system path, and then performs actions on the resulting host-system path.

The new TRUENAME /H option will make it easier to do all this - all I'll need to do is read a text file with the host path, instead of parsing a file to determine that path. But it would be much easier and more reliable if I could give a command like this:

`START doSomething.sh hostname://D:\mydocs\myfile.doc`

That would make it possible to avoid writing a file with the TRUENAME /H path and reading that file before performing the action.

I realize that this may be more trouble than it is worth, but I think that other people might find it useful also. And of course DOSBox-X knows what the host-system path of any file is, so I hope there might be a simple way of entering it.

I thought of trying to enter the host-system path as a variable, but WordPerfect doesn't know how to expand environment variables, so this would be impractical.

**Comment by Wengier** [comment]
@emendelson I have added an option to automatically translate all paths in the command-line to host system paths when launching host system programs (with START command or directly on Windows). It can be set with either the ``starttranspath`` config option or from the drop-down menu ("DOS" => "Host system applications" => "Translate paths to host system paths"). When the option is enabled DOSBox-X will translate all such paths (either relative or absolute paths) when launching host system programs, as long as the file or directory can be found on the host drive. So in this case suppose D:\ is mounted as C:\FILES and the following command:

``
START D:\scripts\doSomething.sh D:\mydocs\myfile.doc
``

Will be translated to something like ``START C:\FILES\scripts\doSomething.sh C:\FILES\mydocs\myfile.doc``.

The updated Windows build is available from:

* [dosbox-x-update.zip](http://cndos.fam.cx/dosware/dosbox-x-update.zip)

Hope this helps.

**Comment by emendelson** [comment]
@Wengier This is superb. It's much better than anything I thought of. Thank you. I am traveling with a Mac until late next week, but I was able to download the current code and will try to build it and test it. I mostly need this feature on the Mac anyway, so this will be perfect. Thank you again.

**Comment by emendelson** [comment]
@Wengier - I've reopened this issue after building the latest code on the Mac and running the SDL version with `startcmd=true` and `starttranspath=true`, as confirmed by `config -get dos`. There are two issues,

First: Under MacOS Big Sur, `TRUENAME /H` works correctly, and returns the hostpath of a file. So, if I enter 

`TRUENAME /H B:\unix\test.sh`

the command returns the full host path of `test.sh`. But the START command does not translate the DOSBox-X path to the host path. This works correctly in my setup:

`START ../../../BDrive/unix/test.sh <parameter>`

But this has no effect:

`START B:\unix\test.sh <parameter>`

Second, as far as I can tell, the new option is not designed to translate filenames used as parameters to a start command. This is where the translation is needed most. For example, It would be good to be able to use this command to start a shell script with a macOS path as an argument to the script:

`START ../../../BDrive/unix/wpd2word.sh C:\WPDOCS\TESTFILE.WPD`

and have the `wpd2word.sh` script start with the parameter `/Users/myname/DOSDisk/WPDOCS/TESTFILE.WPD`.

That is why I hoped there could be something like an URL that would let me give a command like this:

`START ../../../BDrive/unix/wpd2word.sh hostpath://C:\WPDOCS\TESTFILE.WPD`

With the TRUENAME /H command, I can output the host path to a text file and have the script read the text file to find the host path, but there seems to be no way to send the host path to a script as a parameter. Again, I may be the only person in the universe who wants this, but I think it would be useful to anyone using DOSBox-X to perform complex tests.

**Comment by Wengier** [comment]
@emendelson Can you check if the menu item "DOS" => "Host system applications" => "Translate paths to host system paths" has indeed been enabled/checked when you run the START command? If it is not checked, then it means the translation function is not actually enabled. You can always toggle it from the menu, or please check out the latest code for a fix related to the handling of the starttranspath config option.

As for parameters, adding URL style may complicate things ("hostpath" is never a URL protocol I believe), and as I mentioned earlier the new option is designed to translate filenames used as parameters too, not just the the filenames in the command. I have tried your command (START b:\unix\test.sh <parameters> in macOS and it should work. Please try again with the latest code. Thanks!

**Comment by emendelson** [comment]
@Wengier - You were right. "Translate paths to host system paths" was not enabled even though I had enabled `starttranspath` in the conf file. As soon as I turned on that option, everything worked perfectly. I have a slow connection where I am now, so I didn't download current code, but I made the change to dos.cpp that is specified in https://github.com/joncampbell123/dosbox-x/commit/68e168cf6e1d2f188f5ebb4821619f6d2a761987 and built again, and this time, the config option works as described. This is an enormous help. Thank you. I've been removing many lines of code from shell scripts, and simply using "$1" to get the path I need. 

Thank you again, and closing this.
